api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static java.util.Collections.emptySet;
+import static java.util.Collections.singleton;
+
+import com.google.auto.value.AutoValue;
+import com.google.auto.value.extension.memoized.Memoized;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/** javadoc me. */
+@Immutable
+public abstract class Attributes implements Iterable<Entry<String, AttributeValue>> {
```

Comment:
I would go with foreach pattern to avoid allocations.

Reply:
I've been struggling with how to provide this without making some sort of wrapper around the key-value pairs. Do you have an API in mind? Maybe I'm not understanding with you mean by a "foreach" pattern here. I was assuming you wanted to be able to do a java "foreach loop" which looks like this:

```
for(Thing value : attributes) {
}
```

But, what is the "Thing" here?

Reply:
hmm.  maybe a java 8-style `foreach` method that takes a `BiConsumer<String, AttributeValue>` (or equivalent). A little clunky to use in java 7, but very nice in java 8+!